File: core/src/ru/deadsoftware/cavedroid/game/objects/Block.java
package ru.deadsoftware.cavedroid.game.objects;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.math.Rectangle;

import javax.annotation.CheckForNull;

public class Block {

    private final int x;
    private final int y;
    private final int w;
    private final int h;
    private final int hp;
    private final String drop;
    private final String meta;
    @CheckForNull
    private final Sprite tex;

    private final boolean coll;
    private final boolean bg;
    private final boolean tp;
    private final boolean rb;
    private final boolean fluid;

    /**
     * @param left   margin from left edge
     * @param top    margin from top edge
     * @param right  margin from right edge
     * @param bottom margin from bottom edge
     * @param hp     hit points
     * @param drop   id of an item the block will drop when destroyed
     * @param coll   true if block has collision
     * @param bg     true if block should be drawn behind player
     * @param tp     true if block is transparent and renderer should draw a block behind it
     * @param rb     true if block should break when there is no block with collision under it
     * @param fluid  true if fluid
     * @param meta   extra info for storing
     * @param tex    block's texture
     */
    public Block(int left, int top, int right, int bottom, int hp, String drop, boolean coll, boolean bg, boolean tp,
                 boolean rb, boolean fluid, String meta, @CheckForNull Sprite tex) {
        this.x = left;
        this.y = top;
        this.w = 16 - right - left;
        this.h = 16 - top - bottom;
        this.hp = hp;
        this.drop = drop;
        this.coll = coll;
        this.bg = bg;
        this.tp = tp;
        this.rb = rb;
        this.fluid = fluid;
        this.meta = meta;
        this.tex = tex;
        if (this.tex != null) {
            this.tex.flip(false, true);
        }
    }

    public boolean hasCollision() {
        return coll;
    }

    public boolean isBackground() {
        return bg;
    }

    public boolean isTransparent() {
        return tp;
    }

    public boolean requiresBlock() {
        return rb;
    }

    public int getHp() {
        return hp;
    }

    public String getDrop() {
        return drop;
    }

    public boolean hasDrop() {
        return !drop.equals("none");
    }

    public Sprite getTexture() {
        assert tex != null;
        return tex;
    }

    public Rectangle getRectangle(int x, int y) {
        x *= 16;
        y *= 16;
        return new Rectangle(x + this.x, y + this.y, w, h);
    }

    public boolean isFluid() {
        return fluid;
    }

    public String getMeta() {
        return meta;
    }

    public boolean toJump() {
        return (y < 8 && coll);
    }

}

File: core/src/ru/deadsoftware/cavedroid/game/objects/Block.kt
@file:Suppress("DeprecatedCallableAddReplaceWith")

package ru.deadsoftware.cavedroid.game.objects

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.Rectangle
import ru.deadsoftware.cavedroid.game.GameItems

private const val ANIMATION_FRAME_DURATION = 100L
private const val DEPRECATION_MESSAGE =
        "Deprecated since moved to Kotlin. Use generated getter or kotlin property access."

/**
 * @param left          margin from left edge
 * @param top           margin from top edge
 * @param right         margin from right edge
 * @param bottom        margin from bottom edge
 * @param hp            hit points
 * @param drop          id of an item the block will drop when destroyed
 * @param collision     true if block has collision
 * @param background    true if block should be drawn behind player
 * @param transparent   true if block is transparent and renderer should draw a block behind it
 * @param requiresBlock true if block should break when there is no block with collision under it
 * @param fluid         true if fluid
 * @param meta          extra info for storing
 * @param texture       block's texture
 * @param animated      indicates if block has animation
 * @param frames        number of animation frames. ignored if animated is false
 * @param spriteLeft    block's sprite x on texture
 * @param spriteTop     block's sprite y on texture
 * @param spriteRight   block's sprite right edge on texture
 * @param spriteBottom  block's sprite bottom on texture
 */
data class Block(
        val id: Int,
        val left: Int,
        val top: Int,
        val right: Int,
        val bottom: Int,
        val hp: Int,
        val drop: String,
        val collision: Boolean,
        val background: Boolean,
        val transparent: Boolean,
        val requiresBlock: Boolean,
        val fluid: Boolean,
        val meta: String,
        private val texture: Texture?,
        val animated: Boolean,
        val frames: Int,
        private val spriteLeft: Int,
        private val spriteTop: Int,
        private val spriteRight: Int,
        private val spriteBottom: Int
) {

    val width = 16 - right - left
    val height = 16 - top - bottom

    private val spriteWidth = 16 - spriteLeft - spriteRight
    private val spriteHeight = 16 - spriteTop - spriteBottom

    private val sprite: Sprite?
        get() {
            return if (animated) {
                animation[currentFrame()]
            } else {
                field
            }
        }


    private val animation: Array<Sprite>

    init {
        if (frames !in 0..Int.MAX_VALUE) {
            throw IllegalArgumentException("Animation frames must be in range [0, ${Int.MAX_VALUE}]")
        }

        animation = if (animated) {
            if (texture == null) {
                throw IllegalArgumentException("Cannot derive animation frames from null sprite")
            }
            Array(frames) { y ->
                Sprite(texture, spriteLeft, 16 * y + spriteTop, spriteWidth, spriteHeight).apply {
                    flip(false, true)
                }
            }
        } else {
            emptyArray()
        }

        sprite = if (animated) { animation[0] } else {
            if (texture != null) {
                Sprite(texture, spriteLeft, spriteTop, spriteWidth, spriteHeight).apply {
                    flip(false, true)
                }
            } else {
                null
            }
        }
    }

    private fun currentFrame() = if (animated) {
        ((System.currentTimeMillis() / ANIMATION_FRAME_DURATION) % frames).toInt()
    } else {
        0
    }

    fun requireSprite() = sprite ?: throw IllegalStateException("Sprite is null")

    fun draw(spriter: SpriteBatch, x: Float, y: Float) {
        requireSprite().apply {
            setBounds(x + spriteLeft, y + spriteTop, spriteWidth.toFloat(), spriteHeight.toFloat())
            draw(spriter)
        }
    }

    fun getRectangle(x: Int, y: Int) =
            Rectangle(x * 16f + left, y * 16f + this.top, width.toFloat(), height.toFloat())

    fun hasDrop() = drop != "none"

    fun toJump() = top < 8 && collision

    fun getItem() = GameItems.getItem(GameItems.getBlockKey(id))

    @Deprecated(DEPRECATION_MESSAGE)
    fun hasCollision() = collision

    @Deprecated(DEPRECATION_MESSAGE)
    fun isBackground() = background

    @Deprecated(DEPRECATION_MESSAGE)
    fun isTransparent() = transparent

    @Deprecated(DEPRECATION_MESSAGE)
    fun isFluid() = fluid

    @Deprecated(DEPRECATION_MESSAGE)
    fun requiresBlock() = requiresBlock

    @Deprecated("Was renamed to Sprite to comply with variable type.", ReplaceWith("requireSprite()"))
    fun getTexture() = sprite

}